Add extra parameters to FindCommandParser to parse

jevan13 opened the pull request:
## Describe your changes
- Add extra parameters for FindCommandParser.parse()
- Parse using CLI Syntax
- Create separate ContainsKeywordsPredicate classes for simplicity
- Updated Testcases

## Issue ticket number and link
closes #104 

## Checklist before requesting a review
- [X] I have performed a self-review of my code.
- [X] If it is a core feature, I have added thorough tests.

jevan13 committed:
Add extra parameters to FindCommandParser to parse
- Parse using CLI Syntax
- Create separate ContainsKeywordsPredicate classes for simplicity
- Updated Testcases

InfinityTwo commented on src/main/java/seedu/address/logic/commands/FindCommand.java:
```diff
@@ -23,16 +27,16 @@ public class FindCommand extends Command {
             + "Example: " + COMMAND_WORD + " " + ARGUMENT_WORD + " alice bob charlie";
 
-    private final NameContainsKeywordsPredicate predicate;
+    private final PredicateContainer predicateContainer;
```
Perhaps the name declaration here could be changed to SCREAMING_SNAKE_CASE?

InfinityTwo commented on src/main/java/seedu/address/logic/parser/FindCommandParser.java:
```diff
@@ -22,24 +34,67 @@ public class FindCommandParser implements Parser<FindCommand> {
+            RoleContainsKeywordsPredicate roleContainsKeywordsPredicate =
+                    new RoleContainsKeywordsPredicate(roleKeywords);
+            predicateContainer.addRoleContainsKeywordsPredicate(roleContainsKeywordsPredicate);
+        }
```
Perhaps lines 56-90 could be abstracted into a function that dynamically handles the differences? The statements inside every if statement look the same and perhaps the methods for predicateContainer could be abstracted into one function inside the class to handle the input types?

jevan13 replied:
Updated. FindCommandParser will use a PredicateContainer static method to extract the values out from the argumentMultimap. Add more methods in PredicateContainer to set the Predicate fields using the argumentMultimap.

InfinityTwo commented on src/main/java/seedu/address/model/person/DepartmentContainsKeywordsPredicate.java:
```diff
@@ -0,0 +1,45 @@
+ * Tests that a {@code Person}'s {@code Name} matches any of the keywords given.
+ */
+public class DepartmentContainsKeywordsPredicate implements Predicate<Person> {
+    private final List<String> keywords;
```
Ditto on naming constants

InfinityTwo commented on src/main/java/seedu/address/model/person/EmailContainsKeywordsPredicate.java:
```diff
@@ -0,0 +1,44 @@
+ * Tests that a {@code Person}'s {@code Name} matches any of the keywords given.
+ */
+public class EmailContainsKeywordsPredicate implements Predicate<Person> {
+    private final List<String> keywords;
```
Ditto

InfinityTwo commented on src/main/java/seedu/address/model/person/PhoneContainsKeywordsPredicate.java:
```diff
@@ -0,0 +1,44 @@
+ * Tests that a {@code Person}'s {@code Name} matches any of the keywords given.
+ */
+public class PhoneContainsKeywordsPredicate implements Predicate<Person> {
+    private final List<String> keywords;
```
Ditto

InfinityTwo commented on src/main/java/seedu/address/model/person/PredicateContainer.java:
```diff
@@ -0,0 +1,184 @@
+    private DepartmentContainsKeywordsPredicate departmentContainsKeywordsPredicate;
+    private RoleContainsKeywordsPredicate roleContainsKeywordsPredicate;
+
+    public PredicateContainer() {
```
Perhaps there should be an explanation/JavaDoc why the constructor is empty?

jevan13 replied:
Will remove constructor

InfinityTwo commented on src/main/java/seedu/address/model/person/RoleContainsKeywordsPredicate.java:
```diff
@@ -0,0 +1,44 @@
+ * Tests that a {@code Person}'s {@code Name} matches any of the keywords given.
+ */
+public class RoleContainsKeywordsPredicate implements Predicate<Person> {
+    private final List<String> keywords;
```
Ditto

InfinityTwo commented on src/test/java/seedu/address/logic/parser/AddressBookParserTest.java:
```diff
@@ -84,21 +88,44 @@ public void parseCommand_exit() throws Exception {
+        FindPotentialCommand expected3 = new FindPotentialCommand(predicateContainer);
+        String input3 = FindCommand.COMMAND_WORD + " " + FindPotentialCommand.ARGUMENT_WORD + " " + PREFIX_NAME
+                + keywords.stream().collect(Collectors.joining(" "));
+        FindPotentialCommand actual3 = (FindPotentialCommand) parser.parseCommand(input3);
```
I like the comprehensive testing for each Find class but perhaps instead of appending 1, 2, 3 or 4 to the names of the variable, there could be a better naming convention?

InfinityTwo commented on src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:
```diff
@@ -26,23 +31,89 @@ public void parse_invalidArgs_throwsParseException() {
+    public void parse_validNameArgs_returnsFindCommand() {
         // no leading and trailing whitespaces
+        NameContainsKeywordsPredicate nameContainsKeywordsPredicate =
+                new NameContainsKeywordsPredicate(Arrays.asList("Alice", "Bob"));
```
Any reason why you didn't import the valid person's name and/or substring the name to avoid regressions?

InfinityTwo replied:
The same question goes for their details like address, contact, email, etc.

jevan13 replied:
I feel its better for readability, to see the actual inputs, actual and expected results.

InfinityTwo commented on src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:
```diff
@@ -26,23 +31,89 @@ public void parse_invalidArgs_throwsParseException() {
 
     @Test
     public void parse_empty_keywords() {
         assertParseFailure(parser, "all", String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
```
Perhaps you can also import the constant for all, ph and as well?

InfinityTwo commented on src/test/java/seedu/address/model/person/DepartmentContainsKeywordsPredicateTest.java:
```diff
@@ -0,0 +1,94 @@
+        assertFalse(firstPredicate.equals(1));
+
+        // null -> returns false
+        assertFalse(firstPredicate.equals(null));
```
I like the comprehensive test cases but perhaps you could use more more edge case test cases like null and empty strings here for all the other test cases?

jevan13 replied:
What other test cases?

InfinityTwo replied:
Every other test case without null tests

InfinityTwo commented on src/test/java/seedu/address/model/person/NameContainsKeywordsPredicateTest.java:
```diff
@@ -43,19 +43,19 @@ public void equals() {
         // One keyword
         NameContainsKeywordsPredicate predicate = new NameContainsKeywordsPredicate(Collections.singletonList("Alice"));
-        assertTrue(predicate.test(new PersonBuilder().withName("Alice Bob").withIsEmployee(true).build()));
+        assertTrue(predicate.test(new PersonBuilder().withName("Alice").withIsEmployee(true).build()));
```
Ditto for imports

InfinityTwo reviewed (changes requested):
I think your PR generally looks comprehensive but could use less hardcoding of existing constants/test persons. I think the PR is good to go once that, merge conflicts and codecov are fixed.

jevan13 committed:
Add test for FindCommandParser and PredicateContainer

- Ensure correct evaluation of predicates
- Covered edge cases and invalid input handling for both components
- Improved code coverage and ensured functionality aligns with expected behavior

jevan13 committed:
Add import

- Resolve merge conflict

jevan13 committed:
Update import

- Resolve merge conflict

jevan13 committed:
Merge branch 'master' into Add-Extra-Parameters-for-FindCommandParser

# Conflicts:
#	src/test/java/seedu/address/logic/parser/AddressBookParserTest.java
